packages/libp2p/src/dcutr/dcutr.ts

```diff
@@ -0,0 +1,402 @@
+   */
+  async handleIncomingUpgrade (stream: Stream, relayedConnection: Connection): Promise<void> {
+    const options = {
+      signal: AbortSignal.timeout(this.timeout)
+    }
+
+    try {
+      const pb = pbStream(stream, {
+        maxDataLength: MAX_DCUTR_MESSAGE_SIZE
+      }).pb(HolePunch)
+
+      log('A receiving connect')
+      // 3. Upon receiving the Connect, A responds back with a Connect message
+      // containing its observed (and possibly predicted) addresses.
+      const connect = await pb.read(options)
+
+      if (connect.type !== HolePunch.Type.CONNECT) {
+        log('B sent wrong message type')
+        throw new CodeError('DCUtR message type was incorrect', codes.ERR_INVALID_MESSAGE)
+      }
+
+      if (connect.observedAddresses.length === 0) {
+        log('B sent no multiaddrs')
+        throw new CodeError('DCUtR connect message had no multiaddrs', codes.ERR_INVALID_MESSAGE)
+      }
+
+      const multiaddrs = this.getDialableMultiaddrs(connect.observedAddresses)
+
+      if (multiaddrs.length === 0) {
+        log('B had no dialable multiaddrs')
+        throw new CodeError('DCUtR connect message had no dialable multiaddrs', codes.ERR_INVALID_MESSAGE)
+      }
+
+      log('A sending connect')
+      await pb.write({
+        type: HolePunch.Type.CONNECT,
+        observedAddresses: this.addressManager.getAddresses().map(ma => ma.bytes)
+      })
+
+      log('A receiving sync')
+      const sync = await pb.read(options)
+
+      if (sync.type !== HolePunch.Type.SYNC) {
+        throw new CodeError('DCUtR message type was incorrect', codes.ERR_INVALID_MESSAGE)
+      }
+
+      // TODO: when we have a QUIC transport, the dial step is different - for
+      // now we only have tcp support
+      // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
+
+      // Upon receiving the Sync, A immediately dials the address to B
+      log('A dialing', multiaddrs)
+      const connection = await this.connectionManager.openConnection(multiaddrs, {
+        signal: options.signal,
+        priority: DCUTR_DIAL_PRIORITY,
+        force: true
+      })
+
+      log('DCUtR to %p succeeded via %a, closing relayed connection', relayedConnection.remotePeer, connection.remoteAddr)
+      await relayedConnection.close(options)
+    } catch (err: any) {
+      stream.abort(err)
+    } finally {
+      await stream.close(options)
+    }
+  }
+
+  /**
+   * Takes the `addr` and converts it to a Multiaddr if possible
+   */
+  getDialableMultiaddrs (addrs: Array<Uint8Array | string | null | undefined>): Multiaddr[] {
+    const output = []
+
+    for (const addr of addrs) {
+      if (addr == null || addr.length === 0) {
+        continue
+      }
+
+      try {
+        const ma = multiaddr(addr)
+
+        if (!this.isPublicAndDialable(ma)) {
+          continue
+        }
+
+        output.push(ma)
+      } catch {}
+    }
+
+    return output
```

Comment:
The return type of the method is `Multiaddr[]` but the input type is `Array<Uint8Array | string | null | undefined>` so as implemented the suggestion returns `Array<Uint8Array | string | null | undefined>` - e.g. it's missing a `.map` step to turn`Uint8Array | string | null | undefined` into `Multiaddr`  but adding that will involve iterating over the list twice.

Reply:
It's also worth noting that `for..of` is [faster than JS array methods](https://blog.bitsrc.io/finding-the-fastest-loop-type-in-javascript-38af16fe7b4f) since we're not creating new function contexts/stacks/etc for each iteration - it might not be a bottleneck here but `.forEach`, `.filter`, `.map` etc are not free.